staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_test.go

```diff
@@ -352,6 +356,125 @@ func TestCacherWatchSemanticInitialEventsExtended(t *testing.T) {
+		},
+		{
+			name:                       "separate RPC > only etcd gets bookmarks",
+			separateCacheWatchRPC:      true,
+			expectBookmarkOnEtcd:       true,
+			expectBookmarkOnWatchCache: false,
+		},
+		{
+			name:                         "separate RPC & watch cache context > only watch cache gets bookmarks",
+			separateCacheWatchRPC:        true,
+			useWatchCacheContextMetadata: true,
+			expectBookmarkOnEtcd:         false,
+			expectBookmarkOnWatchCache:   true,
+		},
+	}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SeparateCacheWatchRPC, tc.separateCacheWatchRPC)()
+			server, etcdStorage := newEtcdTestStorage(t, etcd3testing.PathPrefix())
+			_, cacher, server, terminate := testSetupCacher(t, server, etcdStorage)
+
+			if err := cacher.ready.wait(context.Background()); err != nil {
+				t.Fatalf("unexpected error waiting for the cache to be ready")
+			}
+			t.Cleanup(terminate)
+
+			requestProgressNotify := func() error {
+				ctx := context.Background()
+				if tc.useWatchCacheContextMetadata {
+					ctx = metadata.NewOutgoingContext(ctx, cacher.watchCache.waitingUntilFresh.contextMetadata)
+				}
+				return etcdStorage.RequestWatchProgress(ctx)
+			}
+
+			gotBookmark := watchCacheIsGettingBookmarks(t, context.Background(), server, cacher, requestProgressNotify)
+			if gotBookmark != tc.expectBookmarkOnWatchCache {
+				t.Errorf("Unexpected watch cache bookmark check result, got: %v, want: %v", gotBookmark, tc.expectBookmarkOnWatchCache)
+			}
+			gotBookmark = etcdWatchIsGettingBookmarks(context.Background(), etcdStorage, requestProgressNotify)
+			if gotBookmark != tc.expectBookmarkOnEtcd {
+				t.Errorf("Unexpected etcd bookmark check result, got: %v, want: %v", gotBookmark, tc.expectBookmarkOnEtcd)
+			}
+		})
+	}
+}
+
+func etcdWatchIsGettingBookmarks(ctx context.Context, etcdStorage storage.Interface, requestProgressNotify func() error) bool {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*2)
+	defer cancel()
+	opts := storage.ListOptions{ResourceVersion: "", Predicate: storage.Everything}
+	opts.Predicate.AllowWatchBookmarks = true
+	w, err := etcdStorage.Watch(ctx, "/pods/", opts)
+	if err != nil {
+		panic(err)
+	}
+	defer w.Stop()
+
+	gotBookmark := false
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for event := range w.ResultChan() {
+			if event.Type == watch.Bookmark {
+				gotBookmark = true
+				break
+			}
+		}
+	}()
+	// Wait to avoid https://github.com/etcd-io/etcd/issues/17507
+	time.Sleep(time.Second)
+	err = requestProgressNotify()
+	if err != nil {
+		panic(err)
+	}
+	wg.Wait()
+	return gotBookmark
+}
+
+func watchCacheIsGettingBookmarks(t *testing.T, ctx context.Context, server *etcd3testing.EtcdTestServer, cacher *Cacher, requestProgressNotify func() error) bool {
+	rv := cacher.resourceVersion()
+	resp, err := server.V3Client.Put(ctx, "foo", "bar")
```

Comment:
Here I want to increase etcd revision in a way that doesn't generate event on watch. As watch cache monitors all events to resource, to get the same result I would need to create a separate etcdStorage on a different resource. 

This would make the test more obscure, as I would need to add a constructor for another resource (`newEtcdTestStorage` hardcodes handling example.Pods) and manage two different etcdStorages. Understand the preference to keep to one abstraction layer, however that also breaks the existing assumption that there is only one etcdStorage in the cacher_tests.

Reply:
There is nothing that prevents you from creating random thing via etcdStorage - you can provide arbitrary key, value there and that's exactly what I was suggesting.

Reply:
Right, just need to figure out how to go through object runtime serialization&deserialization errors.